urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/httpd/resource/HypertextAccessHandler.java

```diff
@@ -0,0 +1,614 @@
+
+	    }
+
+	}
+
+
+	// .htpasswd file exists?
+	if( !authUserFile.exists() ) {
+
+	    throw new ConfigurationException( "Cannot read AuthUserFile: file not found." );
+
+	}
+
+
+
+	// Switch authorization method
+	if( authMethod.equalsIgnoreCase("Basic") ) {
+
+	    return checkValidUser_basicAuthentication( requestURI,
+						       headers,
+						       htaccess,
+						       authMethod,
+
+						       authUserFile,
+
+						       additionalSettings,
+						       optionalReturnSettings
+						       );
+
+	} else if( authMethod.equalsIgnoreCase("Digest") ) {
+
+	    
+	    return checkValidUser_digestAuthentication( requestURI,
+							headers,
+							htaccess,
+							authMethod,
+
+							authUserFile,
+
+							additionalSettings,
+							optionalReturnSettings
+							);
+
+	} else {	    
+
+	    this.logger.log( Level.SEVERE,
+			     getClass().getName() + ".checkValidUser(...)",
+			     "Oooops, the authentication method \""+authMethod+"\" is not supported. Usually this case should be detected much earlier!" );
+	    throw new ConfigurationException( "The authentication method \""+authMethod+"\" is not supported." );
+	}
+	
+    }
+
+    //Constannts.KEY_AUTHORIZATION_DIGESTCHALLENGE
+
+
+    private boolean checkValidUser_basicAuthentication( URI requestURI,
+							HTTPHeaders headers,
+							HypertextAccessFile htaccess,
+							String authMethod,
+							File authUserFile,
+							Map<String,BasicType> additionalSettings,
+							Map<String,BasicType> optionalReturnSettings
+							) 
+	throws IOException,
+	       ConfigurationException {
+
+
+	BasicType wrp_authUser   = additionalSettings.get( Constants.KEY_AUTHORIZATION_USER );
+	BasicType wrp_authPass   = additionalSettings.get( Constants.KEY_AUTHORIZATION_PASS );
+	if( wrp_authUser == null || wrp_authPass == null ) 
+	    return false; // No authorization credentials passed (at least one missing)
+
+	String authUser          = wrp_authUser.getString();
+	char[] authPassData      = wrp_authPass.getString().toCharArray();
+	
+
+	// Locate user in the AuthUserFile
+	char[] filePassData = null;
+	
+	try {
+	    filePassData = HypertextPasswordFile.getPasswordData( authUserFile, 
+								  authUser, 
+								  this.strictMode );
+	    // Found in file?
+	    if( filePassData == null ) 
+		return false;       // Not in the user list
+
+	    // Password data exists -> store into returnSettings
+	    this.storeAuthUserSettings( filePassData, optionalReturnSettings );
+	    
+	    
+	} catch( ParseException e ) {
+
+	    this.logger.log( Level.WARNING,
+			     getClass().getName() + ".checkValidUser_basicAuthentication(...)",
+			     "Cannot verify user. Malformed AuthUserFile: " + e.getMessage() );
+	    throw new ConfigurationException( "Cannot verify user. Malformed AuthUserFile.", e );
+
+	} 
+
+
+ 
+	return compareMD5Passwords( filePassData, authPassData );
+	    
+    }
+
+
+    private boolean checkValidUser_digestAuthentication( URI requestURI,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 320-368 into a separate method.